gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java

```diff
@@ -0,0 +1,132 @@
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.writer;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.instrumented.writer.InstrumentedDataWriterDecorator;
+import gobblin.instrumented.writer.InstrumentedPartitionedDataWriterDecorator;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+import org.apache.avro.SchemaBuilder;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.commons.lang.reflect.ConstructorUtils;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+/**
+ * {@link DataWriter} that partitions data using a partitioner, instantiates appropriate writers, and sends records to
+ * the chosen writer.
+ */
+public class PartitionedDataWriter<S, D> implements DataWriter<D> {
+
+  private static final GenericRecord NON_PARTITIONED_WRITER_KEY =
+      new GenericData.Record(SchemaBuilder.record("Dummy").fields().endRecord());
+
+  private final Optional<WriterPartitioner> partitioner;
+  private final LoadingCache<GenericRecord, DataWriter<D>> partitionWriters;
+  private final Optional<PartitionAwareDataWriterBuilder> builder;
+  private final boolean shouldPartition;
+
+  public PartitionedDataWriter(DataWriterBuilder<S, D> builder, final State state) throws IOException {
+
+    this.partitionWriters = CacheBuilder.newBuilder().build(new CacheLoader<GenericRecord, DataWriter<D>>() {
+      @Override public DataWriter<D> load(final GenericRecord key) throws Exception {
+        return new InstrumentedPartitionedDataWriterDecorator<D>(createPartitionWriter(key), state, key);
+      }
+    });
+
+    if(state.contains(ConfigurationKeys.WRITER_PARTITIONER_CLASS)) {
+      Preconditions.checkArgument(builder instanceof PartitionAwareDataWriterBuilder,
+          String.format("%s was specified but the writer %s does not support partitioning.",
+              ConfigurationKeys.WRITER_PARTITIONER_CLASS, builder.getClass().getCanonicalName()));
+
+      try {
+        this.shouldPartition = true;
+        this.builder = Optional.of(PartitionAwareDataWriterBuilder.class.cast(builder));
+        this.partitioner = Optional.of(
+            WriterPartitioner.class.cast(ConstructorUtils.invokeConstructor(
+                    Class.forName(state.getProp(ConfigurationKeys.WRITER_PARTITIONER_CLASS)), state)));
+        this.builder.get().validatePartitionSchema(this.partitioner.get().partitionSchema());
+      } catch(ReflectiveOperationException roe) {
+        throw new IOException(roe);
+      }
+    } else {
+      this.shouldPartition = false;
```

Comment:
In this design, Gobblin always uses `PartitionedDataWriter`. Users still specify their data writers as usual, and Gobblin transparently handles whether it should partition or not (depending on whether there is a partitioner class or not).